evm_loader/program/src/account_storage/block_hash.rs

```diff
@@ -1,20 +1,112 @@
 use solana_program::slot_history::Slot;
-use std::cmp::Ordering;
+use solana_program::sysvar::slot_hashes::PodSlotHash;
 
-#[must_use]
-pub fn find_slot_hash(value: Slot, slot_hashes_data: &[u8]) -> [u8; 32] {
-    let slot_hashes_len = u64::from_le_bytes(slot_hashes_data[..8].try_into().unwrap());
+use std::{borrow::BorrowMut, cmp::Ordering};
+
+trait SlotHashesProvider {
+    fn fill_slot_hash_slice(&self, data: &mut [u8], sz: usize, offset: usize);
+
+    fn get_slot_hash_slice<const SZ: usize>(&self, offset: usize) -> [u8; SZ] {
+        let mut data: [u8; SZ] = [0; SZ];
+
+        self.fill_slot_hash_slice(&mut data, SZ, offset);
+
+        data
+    }
+
+    const OPTIMIZE_SMALL_BUF: usize = 0;
+}
+
+struct SlotHashesSysvarProvider {}
+
+#[allow(dead_code)]
+fn sol_get_sysvar_stub(
+    _sysvar_id_addr: *const u8,
+    _result: *mut u8,
+    _offset: u64,
+    _length: u64,
+) -> u64 {
+    solana_program::entrypoint::SUCCESS
+}
+
+impl SlotHashesProvider for SlotHashesSysvarProvider {
+    const OPTIMIZE_SMALL_BUF: usize = 32;
+
+    // copy-paste from solana_program::sysvar::get_sys_var which is declared private for some reason
+    fn fill_slot_hash_slice(&self, data: &mut [u8], sz: usize, offset: usize) {
+        let sysvar_id = solana_program::slot_hashes::sysvar::id();
+
+        let sysvar_id: *const u8 = std::ptr::from_ref(&sysvar_id).cast::<u8>();
+
+        let var_addr = std::ptr::from_mut(data).cast::<u8>();
+
+        let sz: u64 = sz.try_into().unwrap();
+        let offset: u64 = offset.try_into().unwrap();
+
+        #[cfg(target_os = "solana")]
```

Comment:
`#[cfg(target_os = "solana")]` can be applied to the entire `SlotHashesSysvarProvider` as it should not be used outside the Solana environment.